src/mlpack/methods/random_forest/random_forest_impl.hpp

```diff
@@ -499,19 +521,28 @@ double RandomForest<
     {
       if (UseDatasetInfo)
       {
-        avgGain += trees[i].Train(bootstrapDataset, datasetInfo,
+        totalGain += tmpTree.Train(bootstrapDataset, datasetInfo,
             bootstrapLabels, numClasses, minimumLeafSize, minimumGainSplit,
             maximumDepth, dimensionSelector);
       }
       else
       {
-        avgGain += trees[i].Train(bootstrapDataset, bootstrapLabels, numClasses,
+        totalGain += tmpTree.Train(bootstrapDataset, bootstrapLabels, numClasses,
             minimumLeafSize, minimumGainSplit, maximumDepth, dimensionSelector);
       }
     }
+
+    // Storing the trained tree at the desired index.
+    if (warmStart)
+      trees[oldNumTrees + i] = tmpTree;
+    else
+      trees[i] = tmpTree;
```

Comment:
This will copy `tmpTree`, which we should probably try to avoid.  In fact, we can avoid `tmpTree` entirely; instead of using `trees[i]` above, you can use `trees[oldNumTrees + i]` regardless of whether it is a warm start, since `oldNumTrees` can be set to 0 when it is a cold start.